r/R/dplyr-funcs-datetime.R

```diff
@@ -101,6 +101,10 @@ register_bindings_datetime <- function() {
     Expression$create("day_of_week", x, options = list(count_from_zero = FALSE, week_start = week_start))
   })
 
+  register_binding("week", function(x) {
+    (call_binding("yday", x) - 1) %/% 7 + 1
+  })
+
```

Comment:
This approach makes sense, although I think there might be a C++ compute function which can do this instead of calculating it manually.

Reply:
I initially used "week" with `first_week_is_fully_in_year=True` but it didn't match and I didn't want to manually adjust that as I feel it uses a different logic than `lubridate::week`. What do you think?

Reply:
I think `lubridate::week` should match `WeekOptions::USDefaults()`. [See here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/api_scalar.h#L421-L425).
If needed we cal always adjust the [week compute function](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc#L413-L478) to also match lubridate.

Reply:
After talking to Rok I am confident it is better to match this PR with "yday" then "week" as `lubirdate::week` is not connected with calander week at all but is a calculation of "complete seven day periods", [see here](https://lubridate.tidyverse.org/reference/week.html). And[ see Arrow C++ week description](https://arrow.apache.org/docs/cpp/compute.html#temporal-component-extraction) for comparison.

As for the PyArrow side, the logic behind [`pc.week()`](https://arrow.apache.org/docs/python/generated/pyarrow.compute.week.html#pyarrow.compute.week) is in line with the week compute function and Pandas [`dt.week`](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.week.html).

What I will try to do is to use a divide C++ function instead of %%.

Reply:
Apparently we had [this discussion](https://github.com/apache/arrow/pull/11026/files#r703012684) before. Sorry I forgot about it.

Reply:
Sounds good @AlenkaF but you I think you should be fine writing things as they were before without wrapping in multiple `call_binding()` functions as the code already converts it into the correct expressions.

Reply:
Sounds good!
Out of curiosity - is there a rule of thumb what to do in such cases? When do you go for multiple calls vs when do you just do it in R?

Reply:
@rok and @AlenkaF Basically, when we're using mathematical operators[1], because `Ops.Expression` has been defined, it means that these are automagically turned into expressions:

https://github.com/apache/arrow/blob/9359026bad4a626de3699e023f96f0c1383d7032/r/R/expression.R#L260-L266

So the rule of thumb would be to wrap them when they're not operators, otherwise just use them as normal.

[1] https://stat.ethz.ch/R-manual/R-devel/library/base/html/groupGeneric.html

Reply:
> After talking to Rok I am confident it is better to match this PR with "yday" then "week" as lubirdate::week is not connected with calander week at all but is a calculation of "complete seven day periods", see here. And see Arrow C++ week description for comparison.

I also had this concern, but indeed like Alenka says _this_ week is literally 7-day periods from January 1. It's almost too deceptively simple compared to ... every other week circumstance 😂 , but that's definitely what it is! Here's the source too: https://github.com/tidyverse/lubridate/blob/7f5323b295813812bf52ffafb1fe26ac8bec95dd/R/accessors-week.r#L27